llvm/lib/Transforms/Scalar/GVNSink.cpp

```diff
@@ -886,8 +919,12 @@ void GVNSink::sinkLastInstruction(ArrayRef<BasicBlock *> Blocks,
 } // end anonymous namespace
 
 PreservedAnalyses GVNSinkPass::run(Function &F, FunctionAnalysisManager &AM) {
-  GVNSink G;
+  auto &DT = AM.getResult<DominatorTreeAnalysis>(F);
+  GVNSink G(&DT);
   if (!G.run(F))
     return PreservedAnalyses::all();
+
+  // PHI nodes get inserted which haven't been added to the Dominator Tree.
```

Comment:
PHI nodes aren't represented in the dominator tree.  The issue is the call to SplitBlockPredecessors().  (You probably could preserve the domtree without much effort; SplitBlockPredecessors() has code to preserve it if you pass in a DomTreeUpdater.)

Reply:
That makes sense. I see the code around `SplitBlockPredecessors`  has no test. And i'm not sure if this is a useful case either. I'm fine removing that part of code altogether. WDYT?

Reply:
That seems like something you could measure pretty easily.

Reply:
nvm. I found a way to update it. Posting the patch soon.